File: analytics/src/test/java/io/appmetrica/analytics/impl/proxy/AppMetricaFacadeProviderTest.java
package io.appmetrica.analytics.impl.proxy;

import android.content.Context;
import android.location.Location;
import android.os.Build;
import io.appmetrica.analytics.impl.AppMetricaFacade;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.MockedStaticRule;
import java.util.Random;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockedStatic;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = Build.VERSION_CODES.LOLLIPOP)
public class AppMetricaFacadeProviderTest extends CommonTest {

    @Mock
    private AppMetricaFacade mAppMetricaFacade;

    private Context mContext;

    private AppMetricaFacadeProvider mProvider;

    @Rule
    public final MockedStaticRule<AppMetricaFacade> sAppMetricaFacade = new MockedStaticRule<>(AppMetricaFacade.class);

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        mContext = RuntimeEnvironment.getApplication();
        mProvider = new AppMetricaFacadeProvider();
    }

    @Test
    public void peekInitializedImpl() {
        when(AppMetricaFacade.peekInstance()).thenReturn(mAppMetricaFacade);
        assertThat(mProvider.peekInitializedImpl()).isEqualTo(mAppMetricaFacade);
    }

    @Test
    public void peekInitializedImplNull() {
        when(AppMetricaFacade.peekInstance()).thenReturn(null);
        assertThat(mProvider.peekInitializedImpl()).isNull();
    }

    @Test
    public void getInitializedImpl() {
        when(AppMetricaFacade.getInstance(mContext, false)).thenReturn(mAppMetricaFacade);
        assertThat(mProvider.getInitializedImpl(mContext)).isEqualTo(mAppMetricaFacade);
    }

    @Test
    public void isInitializedForAppForTrue() {
        when(AppMetricaFacade.isInitializedForApp()).thenReturn(true);
        assertThat(mProvider.isInitializedForApp()).isTrue();
    }

    @Test
    public void isInitializedForAppForFalse() {
        when(AppMetricaFacade.isInitializedForApp()).thenReturn(false);
        assertThat(mProvider.isInitializedForApp()).isFalse();
    }

    @Test
    public void isActivatedTrue() {
        when(AppMetricaFacade.isActivated()).thenReturn(true);
        assertThat(mProvider.isActivated()).isTrue();
    }

    @Test
    public void isActivatedFalse() {
        when(AppMetricaFacade.isActivated()).thenReturn(false);
        assertThat(mProvider.isActivated()).isFalse();
    }

    @Test
    public void markActivated() {
        mProvider.markActivated();
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() {
                AppMetricaFacade.markActivated();
            }
        });
    }

    @Test
    public void setLocation() {
        final Location location = mock(Location.class);
        mProvider.setLocation(location);
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.setLocation(location);
            }
        });
    }

    @Test
    public void setLocationTracking() {
        final boolean locationTracking = new Random().nextBoolean();
        mProvider.setLocationTracking(locationTracking);
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.setLocationTracking(locationTracking);
            }
        });
    }

    @Test
    public void setDataSendingEnabled() {
        final boolean dataSendingEnabled = new Random().nextBoolean();
        mProvider.setDataSendingEnabled(dataSendingEnabled);
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.setDataSendingEnabled(dataSendingEnabled);
            }
        });
    }

    @Test
    public void putErrorEnvironmentValue() {
        final String key = "key";
        final String value = "value";
        mProvider.putErrorEnvironmentValue(key, value);
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.putErrorEnvironmentValue(key, value);
            }
        });
    }

    @Test
    public void putAppEnvironmentValue() {
        final String key = "key";
        final String value = "value";
        mProvider.putAppEnvironmentValue(key, value);
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.putAppEnvironmentValue(key, value);
            }
        });
    }

    @Test
    public void clearAppEnvironment() {
        mProvider.clearAppEnvironment();
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.clearAppEnvironment();
            }
        });
    }

    @Test
    public void setUserProfileID() {
        final String userProfileID = "user_profile_id";
        mProvider.setUserProfileID(userProfileID);
        sAppMetricaFacade.getStaticMock().verify(new MockedStatic.Verification() {
            @Override
            public void apply() throws Throwable {
                AppMetricaFacade.setUserProfileID(userProfileID);
            }
        });
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/proxy/AppMetricaFacadeProviderTest.kt
package io.appmetrica.analytics.impl.proxy

import android.content.Context
import android.location.Location
import io.appmetrica.analytics.impl.AppMetricaFacade
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.staticRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.mockito.Mockito
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.util.Random

internal class AppMetricaFacadeProviderTest : CommonTest() {

    private val appMetricaFacade: AppMetricaFacade = mock()
    private var context: Context = mock()

    @get:Rule
    val appMetricaFacadeStaticRule = staticRule<AppMetricaFacade>()

    private val provider: AppMetricaFacadeProvider by setUp { AppMetricaFacadeProvider() }

    @Test
    fun peekInitializedImpl() {
        whenever(AppMetricaFacade.peekInstance()).thenReturn(appMetricaFacade)
        assertThat(provider.peekInitializedImpl()).isEqualTo(appMetricaFacade)
    }

    @Test
    fun peekInitializedImplNull() {
        whenever(AppMetricaFacade.peekInstance()).thenReturn(null)
        assertThat(provider.peekInitializedImpl()).isNull()
    }

    @Test
    fun getInitializedImpl() {
        whenever(AppMetricaFacade.getInstance(context)).thenReturn(appMetricaFacade)
        assertThat(provider.getInitializedImpl(context)).isEqualTo(appMetricaFacade)
    }

    @Test
    fun isInitializedForAppForTrue() {
        whenever(AppMetricaFacade.isInitializedForApp()).thenReturn(true)
        assertThat(provider.isInitializedForApp).isTrue()
    }

    @Test
    fun isInitializedForAppForFalse() {
        whenever(AppMetricaFacade.isInitializedForApp()).thenReturn(false)
        assertThat(provider.isInitializedForApp).isFalse()
    }

    @Test
    fun isActivatedTrue() {
        whenever(AppMetricaFacade.isActivated()).thenReturn(true)
        assertThat(provider.isActivated).isTrue()
    }

    @Test
    fun isActivatedFalse() {
        whenever(AppMetricaFacade.isActivated()).thenReturn(false)
        assertThat(provider.isActivated).isFalse()
    }

    @Test
    fun markActivated() {
        provider.markActivated()
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.markActivated() }
    }

    @Test
    fun setLocation() {
        val location = Mockito.mock(Location::class.java)
        provider.setLocation(location)
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.setLocation(location) }
    }

    @Test
    fun setLocationTracking() {
        val locationTracking = Random().nextBoolean()
        provider.setLocationTracking(locationTracking)
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.setLocationTracking(locationTracking) }
    }

    @Test
    fun setAdvIdentifiersTracking() {
        provider.setAdvIdentifiersTracking(true)
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.setAdvIdentifiersTracking(true) }
    }

    @Test
    fun setDataSendingEnabled() {
        val dataSendingEnabled = Random().nextBoolean()
        provider.setDataSendingEnabled(dataSendingEnabled)
        appMetricaFacadeStaticRule.staticMock.verify {
            AppMetricaFacade.setDataSendingEnabled(
                dataSendingEnabled
            )
        }
    }

    @Test
    fun putErrorEnvironmentValue() {
        val key = "key"
        val value = "value"
        provider.putErrorEnvironmentValue(key, value)
        appMetricaFacadeStaticRule.staticMock.verify {
            AppMetricaFacade.putErrorEnvironmentValue(
                key,
                value
            )
        }
    }

    @Test
    fun putAppEnvironmentValue() {
        val key = "key"
        val value = "value"
        provider.putAppEnvironmentValue(key, value)
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.putAppEnvironmentValue(key, value) }
    }

    @Test
    fun clearAppEnvironment() {
        provider.clearAppEnvironment()
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.clearAppEnvironment() }
    }

    @Test
    fun setUserProfileID() {
        val userProfileID = "user_profile_id"
        provider.setUserProfileID(userProfileID)
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.setUserProfileID(userProfileID) }
    }

    @Test
    fun addAutoCollectedDataSubscriber() {
        val subscriber = "Subscriber"
        provider.addAutoCollectedDataSubscriber(subscriber)
        appMetricaFacadeStaticRule.staticMock.verify { AppMetricaFacade.addAutoCollectedDataSubscriber(subscriber) }
    }
}
